crates/api/src/config.rs

```diff
@@ -0,0 +1,176 @@
+/// that cannot be changed at runtime, the likes of which might be found
+/// in a configuration file.
+///
+/// If a specific module has a config that can be changed at runtime, the
+/// component found in this type might be a `default_` prefixed version
+/// of it, then the runtime value can be altered through different means.
+///
+/// It is highly recommended that you expose this struct in your module
+/// docs to help devs using your module understand how to configure it.
+pub trait ModConfig:
+    'static
+    + Sized
+    + Default
+    + std::fmt::Debug
+    + serde::Serialize
+    + serde::de::DeserializeOwned
+    + Send
+    + Sync
+{
+}
+
+/// Kitsune configuration.
+#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
+pub struct Config(serde_json::Map<String, serde_json::Value>);
+
+impl Config {
+    /// When kitsune2 is generating a default or example configuration
+    /// file, it will pass a mutable reference of this config struct to
+    /// the module factories that are configured to be used. Those factories
+    /// should call this function any number of times to add any default
+    /// configuration parameters to that file.
+    pub fn add_default_module_config<M: ModConfig>(
+        &mut self,
+        module_name: String,
+    ) -> std::io::Result<()> {
+        if self.0.contains_key(&module_name) {
+            return Err(std::io::Error::other(format!(
+                "Refusing to overwrite conflicting module name: {module_name}"
+            )));
+        }
+        self.0.insert(module_name, tc(&M::default())?);
+        Ok(())
+    }
+
+    /// When kitsune2 is initializing, it will call the factory function
+    /// for all of its modules with an immutable reference to this config
+    /// struct. Each of those modules may choose to call this function
+    /// to extract a module config. Note that this config is loaded from
+    /// disk and can be edited by humans, so the serialization on the module
+    /// config should be tolerant to missing properties, setting sane defaults.
+    /// A module may choose to warn about missing or extraneous properties.
+    pub fn get_module_config<M: ModConfig>(
+        &self,
+        module_name: &str,
+    ) -> std::io::Result<M> {
+        self.0
+            .get(module_name)
+            .map(tc)
+            .unwrap_or_else(|| Ok(M::default()))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn config_usage_example() {
+        #[derive(
+            Debug, Default, serde::Serialize, serde::Deserialize, PartialEq,
+        )]
+        struct Mod1 {
+            #[serde(default)]
+            p_a: u32,
+            #[serde(default)]
+            p_b: String,
+        }
+
+        impl ModConfig for Mod1 {}
+
+        #[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+        #[serde(default)]
+        struct Mod2 {
+            #[serde(rename = "#p_c", skip_deserializing)]
+            doc_p_a: &'static str,
```

Comment:
I'd prefer a JSON schema with a field description. Yes it's a little further from the code but it's usable when editing the document, without having to commit the help text to source control

Reply:
I'm thinking about this from the writing the example/default config file perspective. It's nice if the comments *actually* get written there so humans can have them right next to where they are manually editing the file.

Reply:
Well, this is just an example of how we *could* have it look. I'll not add doc fields like this to any of my config structs until we determine it is desirable.